### PR TITLE
Incremented SFML version number and added changes to the changelog.

### PR DESCRIPTION
**Pull request by eXpl0it3r**
As per usual, here's the "final" commit for SFML 2.4.1

### CODE REVIEW

**Comment by binary1248** [comment]
👍

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

**Comment by jcowgill** [comment]
I believe you are about to break the ABI without changing the SOVERSION. This is the breakage I can see by scanning the header changes. Sorry I only saw this pretty late on.

### Audio
- `sf::InputSoundFile`
 - size increased - not enough memory allocated by old users
- `bool sf::SoundStream::fillAndPushBuffer(unsigned int)`
 - removed symbol

### Graphics
- `virtual bool sf::RenderTexture::activate(bool)`
- `bool sf::RenderWindow::activate(bool)`
 - removed symbols
- `bool sf::RenderWindow::setActive(bool)`
 - new virtual method probably changes the vtable ABI

### Window
- `void sf::GlResource::ensureGlContext()`
 - removed symbol

**Comment by eXpl0it3r** [comment]
Can you explain what negative effects this has, especially for you as packagr maintainer?

(I'm not as familiar with SOVERSION etc)

**Comment by jcowgill** [comment]
This turned out longer than I thought it would, but hopefully covers most stuff!

Any code which uses the affected classes / methods will break if SFML is upgraded. For instance, missing symbols will cause the dynamic linker to simply refuse to load the executable. To fix the programs affected, they need to be rebuilt against the newer SFML.

The usual way for an upstream library project to indicate that it broke the ABI is by changing the SONAME (controlled by CMake's SOVERSION property). The SONAME is embedded into all users of the SFML shared library so that the dynamic linker knows which library to load. It is usually of the format `<library name>.so.<integer>`. Changing the SONAME forces users to recompile their applications, but importantly allows them to continue running against the old library until that's been done - multiple copies of a library can be installed at once as long as their SONAMEs are different.

WRT Debian and Ubuntu specifically - packages using SFML get dependencies automatically added to them which look like this: 'libsfml-window2.4 (>= 2.4)'. Unfortunately this means that if the ABI breaks and the package name stays the same, the dependency isn't strict enough (it should also have a 'strictly less than the ABI broken version' as well). The solution is usually to rename all the library packages. Usually the version part of the SONAME is appended to the package name (the '2.4' in this case). If the ABI breaks and the SONAME does not change, I would have to rename the package to something like 'libsfml-window2.4a'.

Basically this all boils down to "it's frustrating for users and package maintainers to have to rebuild everything / manage library transitions". Correcting the SONAME helps a little and ensures that all package maintainers realize there's been an ABI break.

**Comment by binary1248** [comment]
```diff
diff --git a/cmake/Macros.cmake b/cmake/Macros.cmake
index cd2ca8f..a4b0165 100644
--- a/cmake/Macros.cmake
+++ b/cmake/Macros.cmake
@@ -13,6 +13,9 @@ macro(sfml_add_library target)
     # create the target
     add_library(${target} ${THIS_SOURCES})
 
+    # define the version of the module
+    set_target_properties(${target} PROPERTIES VERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}")
+
     # define the export symbol of the module
     string(REPLACE "-" "_" NAME_UPPER "${target}")
     string(TOUPPER "${NAME_UPPER}" NAME_UPPER)
```
That should do the trick...

**Comment by eXpl0it3r** [comment]
> That should do the trick...

Aren't we doing this already anyways on [line 47](https://github.com/SFML/SFML/blob/master/cmake/Macros.cmake#L43)?

```cpp
    # set the version and soversion of the target (for compatible systems -- mostly Linuxes)
    # except for Android which strips soversion suffixes
    if(NOT SFML_OS_ANDROID)
        set_target_properties(${target} PROPERTIES SOVERSION ${VERSION_MAJOR}.${VERSION_MINOR})
        set_target_properties(${target} PROPERTIES VERSION ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH})
    endif()
```

Question is whether we should adjust the `SOVERSION` as well.

**Comment by binary1248** [comment]
Hmm... yeah I overlooked that part. Seems VERSION was already being set. According to CMake documentation, if we remove the line setting SOVERSION, it should automatically be set to the same value as VERSION, which would indicate a different ABI whenever we bump the patch number.

**Comment by LaurentGomila** [comment]
In my opinion this is less a SOVERSION issue than a violation of our release rules. Patch versions must *not* break the ABI, it must be a drop-in replacement for other releases of the same minor version.

The main issue here is that we completely overlooked that adding a member or changing a virtual function would break the ABI.

File: changelog.txt [hunk 1]
```diff
@@ -1,3 +1,61 @@
...
+
+
+Graphics
+========
+
+Bugfixes
+--------
+* Adjusted uniform error message #1131)
+* Clarify documentation on Rect::contains function bounds (#1151)
+* Added missing setActive virtual method to sf::RenderTarget #1157)
```
Comment (Foaly) [inline changelog.txt]: Tiny typo: there are two opening parenthesis missing in this block.

**Comment by mantognini** [comment]
That nasty ABI thing... :-/

It seems "only" two PRs are mainly responsible for this issue. First, #1118 for the audio part. This one is not about a crash so we could delay it to 2.5 if we can't find a way to avoid adding the extra member.

Second, there's #1157. Here, could we keep `virtual bool RenderTarget::activate(bool)` as the implementation function (private) and add `setActive(bool)` as a proxy to `activate` wherever needed? We could then easily revert this change for 2.5. (I just had a brief look at it, so maybe I'm completely wrong...)

And for `sf::GlResource::ensureGlContext()`, could we keep it with an empty body?

**Comment by LaurentGomila** [comment]
It's clear that PRs that break the ABI should be postponed to the next minor version (SFML 2.5).

A good way to check whether we broke the ABI or not is to see if we touched headers. If so, unless that was to fix typos in comments, we most likely broke it.

**Comment by eXpl0it3r** [comment]
We can easily remove the last three commits from 2.4.x and just keep them in master.

Alternatively we can just remove the last two ones and try and make sure that the changes from the no_internal_context commit aren't breaking the ABI and still include the fix. Which means we'd just have to add an empty body for `ensureContext()` I guess.

**Comment by LaurentGomila** [comment]
> Which means we'd just have to add an empty body for ensureContext() I guess.

Yes, and mark it as deprecated, as usual.

**Comment by eXpl0it3r** [comment]
> mark it as deprecated, as usual.

A bit pointless since it's pretty much an internal function and will be removed with the next minor release, but I can add the deprecation.

**Comment by LaurentGomila** [comment]
> A bit pointless since it's pretty much an internal function and will be removed with the next minor release, but I can add the deprecation.

Yes of course, but strictly speaking, since it is protected, it might be used so we have to explicitly mark it. And most of all, it is a good reminder for the next minor release 😛

**Comment by LaurentGomila** [comment]
Now that I think about it, removing a non-internal function should only be done in the next major release, because otherwise it would break the API.

I guess this one is fine because nobody is likely to use it (and if so, the fix is trivial), but we have to be careful with this stuff.

**Comment by victorlevasseur** [comment]
They never intended to keep the ABI between patch version (even if it's
common in other libraries).

Le ven. 21 oct. 2016 à 22:22, James Cowgill <notifications@github.com> a
écrit :

**Comment by eXpl0it3r** [comment]
@jcowgill is this version of the 2.4.x branch working for you now?

**Comment by jcowgill** [comment]
The audio and graphics modules are fine now.

For window, we need `void sf::GlResource::ensureGlContext()`, not `void sf::priv::GlContext::ensureContext()`. As the latter is in the `priv` namespace, it doesn't affect the ABI.

**Comment by LaurentGomila** [comment]
> As the latter is in the priv namespace, it doesn't affect the ABI.

This is not the case here, but things in the priv namespace may affect the ABI. What they don't affect is the API.

**Comment by binary1248** [comment]
We should really add an issue for 2.5 to move certain things out of the external interface to protect the ABI from internal reorganization when it is necessary, even within patch releases...

**Comment by eXpl0it3r** [comment]
Woops, looks like I was in a bit of a rush when I added that function again. Should be fixed now.

**Comment by eXpl0it3r** [comment]
Had a bit of an issue with pur CI system, thus the delay.

@jcowgill: Can you give it a final look, would be greatly appreciated!

**Comment by jcowgill** [comment]
Everything looks good now. Thanks!

File: changelog.txt [hunk 2]
```diff
@@ -1,3 +1,61 @@
+SFML 2.4.1
+==========
+
+Also available on the website: http://www.sfml-dev.org/changelog.php#sfml-2.4.1
+
+General
+=======
+
+* [kFreeBSD] Define SFML_OS_FREEBSD when compiling for kFreeBSD (#1129
+* [Linux] Correctly add XCB components if no components are requested (#1117)
```
Comment (binary1248) [inline changelog.txt]: This isn't really worth mentioning since we completely got rid of XCB...

File: changelog.txt (hunk 1 above)
Comment (binary1248) [inline changelog.txt]: Didn't this get pushed back to 2.5?

File: changelog.txt [hunk 3]
```diff
@@ -1,3 +1,61 @@
...
+* [Window] Added some simple messaging when trying to build under Cygwin (#1153)
+
+
+Window
+======
+
+Bugfixes
+--------
+* Fixed stack overflow on GlContext creation with multiple threads (#989, #1002)
```
Comment (binary1248) [inline changelog.txt]: Didn't this get pushed back to 2.5?

**Comment by eXpl0it3r** [comment]
Yeah, I noticed that somewhere in the process I pushed the old changelog again. Unfortunately didn't have time to fix it since.

File: changelog.txt (hunk 3 above)
Comment (eXpl0it3r) [inline changelog.txt]: No, this is still included, just with a fix for ABI compatibilities.

File: changelog.txt (hunk 1 above)
Comment (eXpl0it3r) [inline changelog.txt]: Yeah, I noticed that somewhere in the process I pushed the old changelog again. Unfortunately didn't have time to fix it since.

**Comment by eXpl0it3r** [comment]
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.
